python/TestHarness/RaceChecker.py

```diff
@@ -0,0 +1,112 @@
+#* This file is part of the MOOSE framework
+#* https://www.mooseframework.org
+#*
+#* All rights reserved, see COPYRIGHT for full restrictions
+#* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+#*
+#* Licensed under LGPL 2.1, please see LICENSE for details
+#* https://www.gnu.org/licenses/lgpl-2.1.html
+
+import sys
+import itertools
+
+class RaceChecker(object):
+    """docstring for RaceChecker."""
+
+    def __init__(self, all_jobs):
+        """ Create a RaceChecker object that has all the jobs"""
+        self.all_jobs = all_jobs
+
+
+    def findRacePartners(self):
+        """ Collect all the jobs that have race conditions """
+        raceConditionsExist = False
+
+        for job_a, job_b in itertools.combinations(self.all_jobs, 2):
+            for filesInA in job_a.modifiedFiles:
+                for filesInB in job_b.modifiedFiles:
+                    if (filesInA == filesInB):
+                        if not (job_a in job_b.getDownstreamNodes() or job_b in job_a.getDownstreamNodes()):
+                            job_a.racePartners.add(job_b)
+                            job_b.racePartners.add(job_a)
+                            raceConditionsExist = True
+        return raceConditionsExist
+
+    def printRaceConditionsByPrereq(self):
+        """ Print jobs with race conditions that share a prereq """
+
+        colissions = dict()
+        for job in self.all_jobs:
+            if len(job.racePartners) > 0:
+                prereq = str([x.getTestName() for x in job.getUpstreamNodes()])
+                shared = list()
+                shared.append(job.getTestName())
+                for partner in job.racePartners:
+                    if job.getUpstreamNodes() == partner.getUpstreamNodes():
+                        shared.append(partner.getTestName())
+                    colissions[prereq] = shared
+        for prereq in colissions:
+            if prereq != "[]":
+                print("The following share " + prereq + " for their prereq(s)")
+            else:
+                print("The following don't share any prereqs")
+            for job in colissions[prereq]:
+                print(job)
+        return
```

Comment:
Not knowing when/how this Class gets called/used, we have a method you can make use of for printing information. See the following on how you can do that:
https://github.com/idaholab/moose/blob/a8876f99b2ea9983747eb05fde91873e42546524/python/TestHarness/JobDAG.py#L84

It will also help to use the above when RaceChecker is run in parallel; Any print operations will collide with other print operations. This is of course if your RaceChecker can run in parallel. I have not studied your PR close enough to see when it is called.

Reply:
The Class is called once everything is over, however, the printRaceConditionsByPrereq was something I was creating to better organize information that I thought we might want. I never ended up using it, and it doesn't seem to really add information that is different enough or more useful that what the printUniqueRacerSets does, so I'm removing that. Since the information isn't complete until everything has finished, none of this is called until the end when everything stops running in parallel. So I think it's fine?

Reply:
Regardless of when, or how, what ever you add to `job.setOutput()`, will be displayed back to the user, in a nice, formatted way. If you find you are spending any amount of time on "making the stdout look nice", then you should probably be using our print utility.

Reply:
Alright, I've been working on this, but I haven't made any progress yet.  Each time I think I understand it, I run some tests and nothing prints out.  I'll keep working on it.

Reply:
Most likely due to the fact RaceChecker is happening within a thread/child.
Try instead, to set some output using

```python
 job.setOutput("some output")
```

and see what happens.

Reply:
Yeah, I've actually been putting

    job.setOutput("something")

in a lot of places in the code.  I've got some in the TestHarness class, the JobDAG, and in the Job class (where it's self.setOutput) as well as in my RaceChecker class.  I've tried to use the printOutput function in the TestHarness as well as just using the getOutput function from the Job class, but it doesn't seem to be doing anything.

Reply:
One would set job.setOutput() to avoid having to use print statements. The TestHarness calls getOutput at the appropriate time, serially, to print results to the screen. If you are setting output, but not seeing that output, it's possible another class is overwriting that output. As I recall there is some sort of safety setting that prevents this... 

indeed, I see some interesting logic in Job.py: setOutput()

Reply:
So, I can see that setOutput is setting it, because I can call print(getOutput) and it prints it.  I think what's happening is that my code gets called after the TestHarness has already printed everything.  But my code needs to happen after everything, because that's when we know for sure which files have been created/changed.  So how should I address this?  Should I find how TestHarness prints everything and make it happen later?  Or should I just leave print statements?  Or is there another option?